src/SMPCoroutines.java

```diff
@@ -0,0 +1,217 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/*
+* Inspired by
+* "Stable Marriage Problem by Coroutines"
+* Lloyd Allison - 1983
+* */
+public class SMPCoroutines {
+
+    class Man implements Runnable{
+        private int id;
+        private int[] preferences;
+        private int currentProposed;
+        private Woman[] womenList;
+
+        public Man(int id, int[] preferences) {
+            currentProposed = -1;
+            this.id = id;
+            this.preferences = preferences;
+        }
+
+        public void setWomenList(Woman[] womenList) {
+            this.womenList = womenList;
+        }
+
+        @Override
+        public void run() {
+            for (currentProposed = 0; currentProposed < preferences.length; currentProposed++) {
+                   proposeTo(preferences[currentProposed] -1);
+            }
+        }
+
+        private void proposeTo(int womanIndex) {
+            Woman next = womenList[womanIndex];
+            System.out.printf("Man %d proposed to %d\n", id, womanIndex);
+            next.receiveProposal(this.id);
+            try {
+                System.out.printf("Man %d waiting for response.\n", id);
+                synchronized (this) {
```

Comment:
The thing about this algorithm is that men who propose to women are left waiting in case they are rejected. So when a solution is found all men are waiting at this point.